provider/oracle/instance.go

```diff
@@ -0,0 +1,411 @@
+	return o.status
+}
+
+// StorageAttachments returns the storage that was attached in the moment
+// of instance creation. This storage cannot be detached dynamically.
+// this is also needed if you wish to determine the free disk index
+// you can use when attaching a new disk
+func (o *oracleInstance) StorageAttachments() []ociResponse.Storage {
+	o.mutex.Lock()
+	defer o.mutex.Unlock()
+	return o.machine.Storage_attachments
+}
+
+// refresh refreshes the instance raw details from the oracle api
+// this method is mutex protected
+func (o *oracleInstance) refresh() error {
+	o.mutex.Lock()
+	defer o.mutex.Unlock()
+
+	machine, err := o.env.client.InstanceDetails(o.name)
+	// if the request failed for any reason
+	// we should not update the information and
+	// let the old one persist
+	if err != nil {
+		return err
+	}
+
+	o.machine = machine
+	return nil
+}
+
+// waitForMachineStatus will ping the machine status until the timeout
+// duration is reached or an error appeared
+func (o *oracleInstance) waitForMachineStatus(state ociCommon.InstanceState, timeout time.Duration) error {
+	errChan := make(chan error)
+	done := make(chan bool)
+	go func() {
+		for {
+			select {
+			case <-done:
+				return
+			default:
+				err := o.refresh()
+				if err != nil {
+					errChan <- err
+					return
+				}
+				if o.machine.State == ociCommon.StateError {
+					errChan <- errors.Errorf(
+						"Machine %v entered error state",
+						o.machine.Name,
+					)
+					return
+				}
+				if o.machine.State == state {
+					errChan <- nil
+					return
+				}
+				time.Sleep(10 * time.Second)
+			}
+		}
+	}()
+
+	select {
+	case err := <-errChan:
+		return err
+	case <-o.env.clock.After(timeout):
+		done <- true
+		return errors.Errorf(
+			"Timed out waiting for instance to transition from %v to %v",
+			o.machine.State, state,
+		)
+	}
+
+	return nil
+}
+
+// delete will delete the instance and attempt to cleanup any instance related
+// resources
+func (o *oracleInstance) delete(cleanup bool) error {
+	if cleanup {
+		err := o.disassociatePublicIps(true)
+		if err != nil {
+			return err
+		}
+	}
+
+	if err := o.env.client.DeleteInstance(o.name); err != nil {
+		return jujuerrors.Trace(err)
+	}
+
+	if cleanup {
+		// Wait for instance to be deleted. The oracle API does not allow us to
+		// delete a security list if there is still a VM associated with it.
+		iteration := 0
+		for {
+			if instance, err := o.env.client.InstanceDetails(o.name); !oci.IsNotFound(err) {
+				if instance.State == ociCommon.StateError {
+					logger.Warningf("Instance %s entered error state", o.name)
+					break
+				}
+				if iteration >= 30 && instance.State == ociCommon.StateRunning {
+					logger.Warningf("Instance still in running state after %q checks. breaking loop", iteration)
+					break
+				}
+				if oci.IsInternalApi(err) {
+					logger.Errorf("got internal server error from API: %q", err)
+				}
+				time.Sleep(1 * time.Second)
```

Comment:
Please use the environ's clock.Clock here.

`PASS: environ_test.go:236: environSuite.TestStopInstances       50.014s`
it doesn't have to be like this ;)